README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://CRAN.R-project.org/web/checks/check_results_calmate.html"><img border="0" src="https://www.r-pkg.org/badges/version/calmate" alt="CRAN check status"/></a> <a href="https://github.com/HenrikBengtsson/calmate/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/calmate/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>     <a href="https://app.codecov.io/gh/HenrikBengtsson/calmate"><img border="0" src="https://codecov.io/gh/HenrikBengtsson/calmate/branch/develop/graph/badge.svg" alt="Coverage Status"/></a> 
</div>

# calmate: Improved Allele-Specific Copy Number of SNP Microarrays for Downstream Segmentation 

The CalMaTe method calibrates preprocessed allele-specific copy number estimates (ASCNs) from DNA microarrays by controlling for single-nucleotide polymorphism-specific allelic crosstalk. The resulting ASCNs are on average more accurate, which increases the power of segmentation methods for detecting changes between copy number states in tumor studies including copy neutral loss of heterozygosity. CalMaTe applies to any ASCNs regardless of preprocessing method and microarray technology, e.g. Affymetrix and Illumina.


## Citing calmate

Whenever using the **calmate** package or the CalMaTe method, please cite:

* M. Ortiz-Estevez, A. Aramburu, H. Bengtsson, P. Neuvial, and A. Rubio, CalMaTe: A method and software to improve allele-specific copy number of SNP arrays for downstream segmentation, Bioinformatics, Volume 28, Issue 13, 2012, Pages 1793-1794, [doi:10.1093/bioinformatics/bts248](https://doi.org/10.1093/bioinformatics/bts248)

## Installation
R package calmate is available on [CRAN](https://cran.r-project.org/package=calmate) and can be installed in R as:
```r
install.packages("calmate")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/calmate", ref="develop")
```
This will install the package from source.  



NEWS
Package: calmate
================

Version: 0.13.0-9000 [2022-03-08]

 * ...
 

Version: 0.13.0 [2022-03-08]

NEW FEATURES:

 * Setting option 'calmate.troubleshoot' to TRUE will result in an object
   '.calmate.troubleshoot' useful for troubleshooting being saved to the
   global environment if and only if qr.solve() produces an error, e.g.
   singular matrix 'a' in solve.

SOFTWARE QUALITY:

 * Package now requires R (>= 3.2.1) because of aroma.core.

 * Updated package dependencies.

 * SPEEDUP: Utilizing subsetted matrixStats (>= 0.50.0) calculations.

 * Package test coverage is 49% (was 32%).

DOCUMENTATION:

 * Add CITATION file.

BUG FIXES:

 * calmateByThetaAB() and calmateByTotalAndFracB() generated a "too few
   arguments" error when they tried to generate another error.


Version: 0.12.1 [2015-10-26]

SIGNIFICANT CHANGES:

 * Package now requires R (>= 3.1.1) released July 2014. This allows us to
   use Bioconductor (>= 3.0) (October 2014).
  
SOFTWARE QUALITY:

 * Explicitly importing core R functions.


Version: 0.12.0 [2015-02-23]

SOFTWARE QUALITY:

 * Package test coverage is 32%.

 * Updated package dependencies.

BUG FIXES:

 * Added missing import.


Version: 0.11.2 [2014-01-30]

SOFTWARE QUALITY:

 * Updated package dependencies.


Version: 0.11.1 [2013-10-17]

DOCUMENTATION:

 * Updated the reference to the CalMaTe paper.


Version: 0.11.0 [2013-10-17]

SIGNIFICANT CHANGES:

 * Package now requires R (>= 2.15.0) [and Bioc (>= 2.10.0)].

SOFTWARE QUALITY:

 * Now importing only what needs to be imported and formally declaring all
   S3 methods in NAMESPACE.

 * Updated package dependencies.


Version: 0.10.3 [2013-09-21]

SOFTWARE QUALITY:

 * Now using only 'Authors@R' in DESCRIPTION available since R (>= 2.14.0).

 * Updated package dependencies.


Version: 0.10.2 [2013-08-03]

SOFTWARE QUALITY:

 * Updated examples and system tests to use matrixStats::rowMedians() so that
   they work with aroma.core (>= 2.10.0) without having to load the
   'matrixStats' package.

 * Added an Authors@R field to DESCRIPTION.

 * Updated package dependencies.

BUG FIXES:

 * Forgot to import rowAlls() from 'matrixStats'.


Version: 0.10.1 [2013-05-22]

CRAN POLICY:

 * Now all Rd \usage{} lines are at most 90 characters long.

SOFTWARE QUALITY:

 * Updated package dependencies.


Version: 0.10.0 [2013-01-05]

NEW FEATURES:

 * Now CalMaTeCalibration implements ParametersInterface.

SOFTWARE QUALITY:

 * Updated package dependencies.

 * Now package declares S3 methods in the namespace.

 * CLEANUP: Now packages MASS, matrixStats, R.methodsS3 and R.oo are only
   imported (not loaded/attached).

BUG FIXES:

 * Previous version was not compatible with R.filesets 1.9.0 and aroma.core
   2.8.0 resulting in "Error: 'sapply' is not an exported object from
   'namespace:R.filesets'".  Thanks to Zheng Siyuan (MD Anderson) for reporting
   on this.


Version: 0.9.3 [2012-04-24]

NEW FEATURES:

 * Updated package dependencies.

DOCUMENTATION:

 * Updated references to the CalMaTe manuscript.


Version: 0.9.2 [2012-03-18]

DOCUMENTATION:

 * Updated help("calmate").


Version: 0.9.1 [2012-02-23]

SOFTWARE QUALITY:

 * Now package explicitly imports 'utils' and no longer relies on 'R.oo'
   and/or 'R.utils' for loading 'utils'.


Version: 0.9.0 [2012-02-21]

SIGNIFICANT CHANGES:

 * CalMaTe algorithm v2 is being finalized. Main updates are:
 
   (a) The estimation of the model parameters are now done solely based on
       reference samples.  In previous versions, some of the initial estimation
       steps were using also non-reference samples.
       
   (b) For a small number of SNPs, the main CalMaTe scheme for estimating
       parameters would not converge or converge poorly.  For such SNPs,
       CalMaTe now falls back to using a median-based estimator.  For more
       details, see help(fitCalMaTeInternal).


Version: 0.8.1 [2012-02-19]

SIGNIFICANT CHANGES:

 * Made findUnitsTodo() for CalMaTeCalibration smarter. Before it would detect
   all non-polymorphic loci as non-fitted.


Version: 0.8.0 [2012-02-19]

NEW FEATURES:

 * BACKWARD COMPATIBILITY: Added argument 'flavor' to calmateByThetaAB() and
   CalMaTeCalibration so that it is possible use previous versions of CalMaTe
   model estimators.  The flavor is now reflected in the "asterisk" (the
   default) tags of CalMaTeCalibration.

SOFTWARE QUALITY:

 * SPEEDUP: The internal CalMaTe fit function is now called directly, which
   avoids the method dispatch overhead that otherwise applies to each SNP
   fitted.

 * ROBUSTNESS: Added system tests to test all flavors of the fitting algorithms.


Version: 0.7.4 [2012-02-06]

NEW FEATURES:

 * Now additional arguments '...' of CalMaTeCalibration(...) are passed to
   calmateByTotalAndFracB() by process().  This makes it possible to specify
   all tuning parameters of the CalMaTe method.

DEPRECATED AND DEFUNCT:

 * Now argument 'references' of process() is obsolete and gives an error.
   Specify it via CalMaTeCalibration() instead.


Version: 0.7.3 [2012-01-31]

 * Updated package dependencies.

BUG FIXES:

 * The bug fix in v0.7.2 for handling the special case where a SNP has the same
   genotype in all samples introduced another bug giving incorrect results in
   that special case.


Version: 0.7.2 [2011-11-15]

NEW FEATURES:

 * ROBUSTNESS: Improved validation of argument 'references' to
   CalMaTeCalibration(), calmateByTotalAndFracB(), and calmateByThetaAB().

DOCUMENTATION:

 * Added a section 'Reference samples' to the help of CalMaTeCalibration.

DEPRECATED AND DEFUNCT:

 * Argument 'references' of process() of CalMaTeCalibration is deprecated.
   Instead, pass it to CalMaTeCalibration().

BUG FIXES:

 * If in some SNPs, the samples include only a single genotype (usually AA or
   BB), the genotype AA is interpreted as BB in half of the samples and vice
   verse (and the returned value is reversed again). There was a bug in the

R/fitCalMaTeV2.R
fitCalMaTeV2 <- function(dataT, references, fB1=1/3, fB2=2/3, maxIter=50, ...) {
  # This is an internal function. Because of this, we will assume that
  # all arguments are valid and correct.  No validation will be done.
  dataInit <- dataT;
  nbrOfSNPs <- nrow(dataT);
  nbrOfReferences <- length(references);

  # Adding a small value so there are "non" 0 values
  eps <- 1e-6;
  dataT[dataT < eps] <- eps;

  eps2 <- 1e-4;
  a <- max(max(dataT[2,references] / (pmax(dataT[1,references],0) + eps2)), max(dataT[1,references] / (pmax(dataT[2,references],0) + eps2)));
  Giro <- matrix(c(1, 1/a, 1/a, 1), nrow=2, ncol=2, byrow=FALSE);
  Giro <- solve(Giro);
  dataT <- Giro %*% dataT;

  # Extract the signals for the reference set
  TR <- dataT[,references, drop=FALSE];

  # Set some weights based on the median
  S <- colSums(TR);
  S[S < 0] <- 0;
  CN <- 2*S/median(S);

  # The use of sqrt(sqrt(...)) is intended. /AR 2012-01-31
  w1 <- 0.1  + 0.9 * sqrt(sqrt(2*(1-pnorm(abs(CN-2) / median(abs(CN-2))))));

  if (sum(is.nan(w1)) > 0) {
    w1 = rep(1, times=length(w1));
  }


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Checking if all the samples are homozygous
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  fracB <- TR[2,] / (TR[1,] + TR[2,]);
  naiveGenoDiff <- 2*(fracB < fB1) - 2*(fracB > fB2);


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Twist half of the reference samples in case there is only one allele?
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  onlyOneAllele <- (abs(sum(naiveGenoDiff)/2) == length(naiveGenoDiff));
  if (onlyOneAllele) {
    idxsSwap <- references[seq_len(ncol(TR)/2)];
    dataT[1:2,idxsSwap] <- dataT[2:1,idxsSwap, drop=FALSE];

    # Update precalcalculated signals
    TR <- dataT[,references, drop=FALSE];
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Total copy numbers must be close to 2 for the reference samples or
  # (if there are not control samples) for most of the samples
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  H <- matrix(2, nrow=nbrOfReferences, ncol=1, byrow=FALSE);
  # Alternatively to the below, on can use method = "MM". /AR 2011-12-04
  suppressWarnings({
    fit <- rlm(H ~ 0 + t(TR), maxit=maxIter, weights=w1);
  });

  # Use fallback estimator, iff failed to converge
  if (!fit$converged) {
    return(fitCalMaTeMedians(dataInit, references=references, fB1=fB1, fB2=fB2));
  }

  matSum <- fit$coefficients;
  coeffs <- fit$w;
  eps3 <- 1e-8;
  coeffs[coeffs < eps3] <- eps3;
  coeffs <- coeffs * w1;
  dataT <- diag(matSum) %*% dataT;

  # Reextract the signals for the reference set
  TR <- dataT[,references, drop=FALSE];

  # The difference of the copy numbers must be 2, 0 or -2 depending genotyping
  fracB <- TR[2,] / (TR[1,] + TR[2,]);
  naiveGenoDiff <- 2*(fracB < fB1) - 2*(fracB > fB2);

  # If all samples are called homozygous, use the fallback estimator,
  # because rlm() is very sensitive to such setups.
  if (abs(sum(naiveGenoDiff)/2) == length(naiveGenoDiff)) {
    return(fitCalMaTeMedians(dataInit, references=references, fB1=fB1, fB2=fB2));
  }

  suppressWarnings({
    fit <- rlm(naiveGenoDiff ~ 0 + t(TR), maxit=maxIter, weights=coeffs);
  });

  # Use fallback estimator, iff failed to converge
  if (!fit$converged) {
    return(fitCalMaTeMedians(dataInit, references=references, fB1=fB1, fB2=fB2));
  }

  matDiff <- fit$coefficients;

  # T matrix is:
  #  [1  1] [   ] = [MatSum[1]   MatSum[2]] (We have already applied it) MatSum is 1,1
  #  [1 -1] [ T ]   [MatDiff[1] MatDiff[2]]
  U <- matrix(c(0.5, 0.5, 0.5, -0.5), nrow=2, ncol=2, byrow=FALSE);
  V <- matrix(c(c(1, 1), matDiff), nrow=2, ncol=2, byrow=TRUE);
  T <- U %*% V;

  res <- T %*% dataT;

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Undo the previous change applied to the data in case there is
  # only one allele
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (onlyOneAllele) {
    res[1:2,idxsSwap] <- res[2:1,idxsSwap, drop=FALSE];
  }

  res;
} # fitCalMaTeV2()


###########################################################################
# HISTORY:
# 2012-02-21 [HB]
# o Now passing arguments 'fB1' and 'fB2' to fitCalMaTeMedians().
# o CODE CLEANUP: Further code cleanup of the new code.
# o Redid the the below code cleanup done Jan 1-Feb 19, 2012.
# 2012-02-21 [MO]
# o Created from fitCalMaTeV1. [HB?!?]
# o Use of weigths on "rlm" function.
# o Warnings by rlm(), which are often due to non-convergence, are now
#   suppressed, because we now take care of that case via the fallback
#   estimator fitCalMaTeMedians().
# o Use of fitCalMaTeMedians() when rlm() does not converge.
# 2012-02-20 [HB]
# o Extracted plain function fitCalMaTeV2() from former fitCalMaTe()
#   for matrices.
# 2012-02-19 [HB]
# o Clarified in the source code comments that it is only the reference
#   samples that are "twisted".
# o Created internal fit functions for the different versions of CalMaTe.
# 2012-01-31 [HB]
# o CLEANUP: Pruning up the code, e.g. dropping extraneous spaces and
#   parenthesis and adding missing ones. Dropping ambigous comments
#   using terminologies such as "previous", "latest" and "final".
# o DOCUMENTATION: Argument "references" have to be an index vector
#   (and cannot be a logical vector as previously said).
# 2012-01-31 [MO]
# o BUG FIX: the index "idxs" was recalculated to undo the change when
#   there is only one allele, and it was done as the previous version,
#   taking into account all the samples, not only the references.
# 2011-12-15 [HB]
# o CORRECTION: Now doing x[x < eps] <- eps instead of x[x < 0] <- eps.
# o Dropped the validation of argument 'references'.  This is an internal
#   function and validation should already have been done elsewhere.
# 2011-12-07 [MO]
# o At least 3 reference samples.
# 2011-12-04 [AR]
# o BUG FIX: there was a bug for SNPs with a single allele when using a
#   set of references.
# o Set some initial weights based on the median that improves the
#   breakdown point if no reference samples are provided when using a set
#   of references.
# 2011-11-29 [MO]
# o Change matrix "T" by "dataT" and "P" by "T"
# 2010-08-02 [HB]
# o ROBUSTNESS: Now fitCalMaTe() also works (technically) when there is
#   only one reference.
# o Made into an S3 method for matrix:es.
# 2010-06-18 [HB]
# o Created from refineCN.list().
###########################################################################
